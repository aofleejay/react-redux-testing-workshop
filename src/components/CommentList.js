import React from 'react'
import CommentItem from './CommentItem'

const CommentList = ({ comments, loadMore }) => (
  <div>
    <div>
      {
        comments.map(({ id, title, body, author, postat }) => (
          <CommentItem
            key={id}
            title={title}
            body={body}
            author={author}
            postat={postat}
          />
        ))
      }
    </div>
    <button
      id="load-more"
      className="button is-primary load-more"
      onClick={loadMore}
    >
      Load More
    </button>
  </div>
)

export default CommentList
