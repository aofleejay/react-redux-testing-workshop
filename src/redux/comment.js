import axios from 'axios'

const GET_COMMENTS_SUCCESS = 'GET_COMMENTS_SUCCESS'
const GET_COMMENTS_FAILED = 'GET_COMMENTS_FAILED'

const initialState = []

const commentReducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_COMMENTS_SUCCESS:
      return [...state, ...action.comments]

    case GET_COMMENTS_FAILED:
      return state

    default:
      return state
  }
}

const receiveComments = comments => ({
  type: GET_COMMENTS_SUCCESS,
  comments: comments,
})

const getComments = () => dispatch =>
  axios
    .get('http://localhost:3000/comments')
    .then(response => response.data)
    .then(comments => dispatch(receiveComments(comments)))
    .catch(() => dispatch({ type: GET_COMMENTS_FAILED }))

export { receiveComments, getComments }
export default commentReducer
