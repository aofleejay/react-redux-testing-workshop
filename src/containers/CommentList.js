import React, { Component } from 'react'
import { connect } from 'react-redux'
import CommentList from '../components/CommentList'
import { getComments } from '../redux/comment'

export class CommentListContainer extends Component {
  componentDidMount() {
    this.props.getComments()
  }

  render() {
    return (
      <CommentList 
        comments={this.props.comments}
        loadMore={this.props.getComments}
      />
    )
  }
}

export const mapStateToProps = ({ comments }) => ({ comments })

export const mapDispatchToProps = dispatch => ({
  getComments: () => dispatch(getComments()),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CommentListContainer)
