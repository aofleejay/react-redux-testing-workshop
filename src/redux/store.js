import { createStore, applyMiddleware, combineReducers } from 'redux'
import thunk from 'redux-thunk'
import commentReducer from './comment'

const reducers = combineReducers({
  comments: commentReducer,
})

const store = createStore(reducers, applyMiddleware(thunk))

export default store
