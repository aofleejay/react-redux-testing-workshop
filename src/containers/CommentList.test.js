import React from 'react'
import { shallow } from 'enzyme'
import { CommentListContainer, mapDispatchToProps, mapStateToProps } from './CommentList'
import * as actions from '../redux/comment'

jest.mock('../components/CommentList', () => 'CommentList')
jest.mock('../redux/comment')

describe('Test CommentList container', () => {
  const props = {
    comments: [
      { id: 1, title: 'title1', body: 'body1', author: 'author1', postat: '1m ago' },
      { id: 2, title: 'title2', body: 'body2', author: 'author2', postat: '2m ago' },
    ],
    getComments: () => {},
  }

  it('Should match its snapshot', () => {
    const wrapper = shallow(<CommentListContainer {...props} />, { disableLifecycleMethods: true })

    expect(wrapper).toMatchSnapshot()
  })

  describe('Test componentDidMount', () => {
    afterEach(() => {
      jest.restoreAllMocks()
    })

    it('Should call getComments once', () => {
      props.getComments = jest.fn()
      const instance = shallow(<CommentListContainer {...props} />, { disableLifecycleMethods: true }).instance()
  
      instance.componentDidMount()
    
      expect(props.getComments).toHaveBeenCalledTimes(1)
    })
  })
  
  describe('Test mapStateToProps function', () => {
    it('Should map comments state to prop', () => {
      const state = { 
        comments: [
          { title: 'title1', body: 'body1', author: 'author1', postat: '1m' },
          { title: 'title2', body: 'body2', author: 'author2', postat: '2m' },
        ],
      }

      const subscribeState = mapStateToProps(state)

      expect(subscribeState).toMatchSnapshot()
    })
  })
  
  describe('Test mapDispatchToProps function',() => {
    it('Shouldmap dispatch to props correctly', () => {
      const spyDispatch = jest.fn()
      const { getComments } = mapDispatchToProps(spyDispatch)
      
      getComments()

      expect(spyDispatch).toHaveBeenCalledTimes(1)
      expect(actions.getComments).toHaveBeenCalledTimes(1)
    })
  })
})