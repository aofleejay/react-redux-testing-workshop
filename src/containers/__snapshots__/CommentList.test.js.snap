// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CommentList container Should match its snapshot 1`] = `
<CommentList
  comments={
    Array [
      Object {
        "author": "author1",
        "body": "body1",
        "id": 1,
        "postat": "1m ago",
        "title": "title1",
      },
      Object {
        "author": "author2",
        "body": "body2",
        "id": 2,
        "postat": "2m ago",
        "title": "title2",
      },
    ]
  }
  loadMore={[Function]}
/>
`;

exports[`Test CommentList container Test mapStateToProps function Should map comments state to prop 1`] = `
Object {
  "comments": Array [
    Object {
      "author": "author1",
      "body": "body1",
      "postat": "1m",
      "title": "title1",
    },
    Object {
      "author": "author2",
      "body": "body2",
      "postat": "2m",
      "title": "title2",
    },
  ],
}
`;
